### ISSUE TITLE
Attributes for step definitions and hooks

### ISSUE THREAD

**Issue by ciaranmcnulty** [Behat/Behat#1300]
PHP 8.0 will have Attributes (https://wiki.php.net/rfc/attributes_v2)

A natural way to use these in Behat is for step definitions.

Before:
```php
/**
 * @beforeScenario
 */
function getStuffReady()
{
   //...
}

/**
 * @Given :user has an idea to :idea
 */
function userHasAnIdeaTo($user, $idea)
{
    //...
}
```

After:
```php
<< BeforeScenario >>
function getStuffReady()
{
   //...
}

<< Given(':user has an idea to :idea') >>
function userHasAnIdeaTo($user, $idea)
{
    //...
}
```

This would be slightly less fragile than the current regex approach (although to be fair it's never really bitten me)

Because the attributes will have an FCQN, we could support 'different' step definition classes for different types of pattern (e.g. Gherkin vs Turnip vs future syntaxes).

There are other places in the code we may want to use attributes, not least some of the empty 'tagging' interfaces like `SnippetAcceptingContext`

**Comment by ciaranmcnulty** [comment]
There would be some added complexity when generating snippets because we would need to `use` the classes at head of file

**Comment by rpkamp** [comment]
I've been looking into this for a bit, and my first thought would be:

1. Introduce `Attribute` classes for the necessary elements (Given, When, Then, BeforeScenario, etc, etc). Using e.g. the call definitions directly would couple them too much to attributes and would need the properties to be public - these properties are now private so that would be a BC break.
2. Introduce a `Behat\Behat\Context\Reader\AttributeContextReader`, similar in setup to `Behat\Behat\Context\Reader\AttributeContextReader` but looking at Attributes rather then at annotations.
3. Introduce a `Behat\Behat\Context\Reader\CompositeContextReader` that combines the results from multiple `ContextReader`s. Specifically here the results of the `AnnotationContextReader` as well as the `AttributeContextReader`. 
4. Register this `CompositeContextReader` in the DI container instead of the `AnnotationContextReader` so that can read both annotations and attributes. 

The `CompositeContextReader` has the advantage that at some in the (probably far away) future the `AnnotationContextReader` and reader can be removed and then the `AttributeContextReader` can be registered as reader directly. At that point the `CompositeContextReader` can be removed from the code too. 

An open question that needs to be handled here: what to do when some method has an attribute _as well as_ an annotation. Will it just accept both? Does need to remove duplicates? Does it need to warn/error when the annotation and attribute say different things about the same method?

**Comment by ciaranmcnulty** [comment]
> what to do when some method has an attribute as well as an annotation

The main reason I can see for this situation arising (aside from forgetting to remove the docblocks) is in a codebase that supports 7.x and 8.x in parallel. In that case maintainers may want to have both attributes and docblocks, and rely on test failures to detect when the two are not in sync.

> Does need to remove duplicates?

We'd need to handle it explicitly somehow. The current code has a runtime error if a step can't be matched to exactly one matcher (each step is run against every matcher every time). This is meant to guard against things like:

```php
/**
 * @Given /I have a .*/
 */
function foo(){}

/**
 * @Given I have a cat
 */
function foo(){}
```

but also has the effect of currently disallowing:

```php
/**
 * @Given I have a cat
 * @Given I have a cat
 */
function foo(){}
```

The simplest solution is probably that if the attributes are present, the docblock isn't read

**Comment by ciaranmcnulty** [comment]
Having considered more, I guess we could change the logic to allow duplicate steps at runtime if they:

 1. match the same method
 2. result in the same captured arguments
 
 I think then we'd be able to simply register all the docblocks + all the attributes without disambiguating
 
 This would then be fine:

```php
/**
 * @Given I have :count bananas
 */
#[Given( 'I have :count bananas' )]
```

**Comment by rpkamp** [comment]
> The current code has a runtime error if a step can't be matched to exactly one matcher (each step is run against every matcher every time)

That seems to be "further down the line" than the annotation readers. Which it should be, it's a separate concern. 

So it seems to me if we just read Attributes and convert them to Call definitions they'll be sent down the same path as the the Call definitions read for Annotations and the whole reading of Attributes is just an infrastructural concern that doesn't need concern itself with those issues at all. 

I'll see if I can take a stab at this. Seems fun 🙂

**Comment by stof** [comment]
Well, to me, we should just register both definitions, just like we support using multiple annotations on the same method.

I don't think we need to do a special case for the case where you register steps both with annotations and attributes at the same time for the *same* step definitions. I don't see any need for such complexity: if you need to support PHP 7, use the annotation way. That *also* works with PHP 8. Duplicating the registration does not simplify anything for context authors (as they now need to keep the duplicates in sync)

**Comment by rpkamp** [comment]
Makes perfect sense to me 👍

**Comment by pamil** [comment]
Should be all covered in v3.9.0 and v3.10.0 thanks to @rpkamp's PRs 🥇

**Comment by rpkamp** [comment]
@pamil What hasn't been covered yet is code generation. That will still generate new steps using annotations rather than attributes. 

But maybe that would be better off in a new issue?

**Comment by ciaranmcnulty** [comment]
That probably deserves a separate issue

**Comment by pamil** [comment]
@rpkamp good catch, we could handle that in a separate issue indeed

**Comment by rpkamp** [comment]
I've created an issue: https://github.com/Behat/Behat/issues/1379
